liquibase/liquibase
# look for jars in local ./liquibase_libs

## Environment
CLI executable files for Windows, MacOS and Linux

**Liquibase Version**: 4.4.1

**Liquibase Integration & Version**: CLI

**Liquibase Extension(s) & Version**: N/A

**Database Vendor & Version**: N/A

**Operating System Type & Version**: MacOS, Linux, and Windows

## Pull Request Type
* \[ ] Bug fix (non-breaking change which fixes an issue.)
* \[x] Enhancement/New feature (non-breaking change which adds functionality)
* \[ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description
When running liquibase in the cli, the executables generate a classpath with jars from LIQUIBASE_HOME and LIQUIBASE_HOME/lib. This PR loads an additional directory of `liquibase_libs` in the current working directory to the classpath. This would allow for individual projects to have separate extensions, drivers, and versions specific to that project (similar to how npm handles this with the node_modules directory).

## Steps To Reproduce
* Create a new directory and add a simple change log file:

```java
<?xml version="1.0" encoding="UTF-8"?>
<databaseChangeLog
        xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
         http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-4.3.xsd">
    <changeSet id="1" author="Liquibase">
        <createTable tableName="test_table">
            <column name="test_id" type="int">
                <constraints primaryKey="true"/>
            </column>
            <column name="test_column" type="varchar"/>
        </createTable>
    </changeSet>
</databaseChangeLog>
```

* Add `driver: org.postgresql.Driver` to your liquibase.properties file. 
* Run `mkdir liquibase_libs`
* Run `cd liquibase_libs && curl https://repo1.maven.org/maven2/org/postgresql/postgresql/42.2.23/postgresql-42.2.23.jar -L -O -J`
* Run `liquibase update --changeLogFile=changelog.xml`

## Actual Behavior
* Expect `Unexpected error running Liquibase: java.lang.RuntimeException: Cannot find database driver: org.postgresql.Driver`

## Expected/Desired Behavior
* Expect success update with driver found in `liquibase_libs` directory.

## Fast Track PR Acceptance Checklist:


<

!--- If you're unsure about any of these, just ask us in a comment. We're here to help|width=200,height=183!

-->

* \[ ] Build is successful and all new and existing tests pass
* \[ ] Added [Unit Test(s)](https://liquibase.jira.com/wiki/spaces/LB/pages/1274937609/How+to+Write+Liquibase+Core+Unit+Tests)
* \[ ] Added [Integration Test(s)](https://liquibase.jira.com/wiki/spaces/LB/pages/1276608569/How+to+Write+Liquibase+Core+Integration+Tests)
* \[ ] Added [Test Harness Test(s)](https://github.com/liquibase/liquibase-test-harness/pulls)
* \[ ] Documentation Updated

## Need Help?
Come chat with us on our [discord channel](https://discord.com/channels/700506481111597066/700506481572839505)

- - -
## Dev Handoff Notes (Internal Use)
#### Links
* Fixed Issue: Stand alone PR
* Test Results: [https://github.com/liquibase/liquibase/pull/2015/checks](https://github.com/liquibase/liquibase/pull/2015/checks)

#### Testing
* Steps to Reproduce: See above
* Guidance:
  * Only impacts CLI launcher code. No java code changed

#### Dev Verification
Only reviewed code

#### Github Users
* mcred

## Test Requirements (Internal Liquibase QA)
**Manual Tests**

Note that the liquibase_libs directory will not exist by default. 

The name of the liquibase_libs directory may change between writing these test cases and the actual testing of them. Check with your developer for the final name.

**Case 1**

_Verify Liquibase CLI launches without error when the /liquibase_libs directory doesn’t exist._

* liquibase --version is successful
* liquibase update --changelog-file <mychangelog> is successful.

**Case 2**

Setup

* If there is a `classpath` in your liquibase.properties file, remove the driver from the definition. For example, mine  `classpath:postgresql-42.2.18.jar;C:\\`. I would need to delete everything up to and including the `;`.

_Verify Liquibase finds the database driver when the driver is in the /lib directory and the driver is not specified in the classpath._

* liquibase update --changelog-file <mychangelog> is successful.

**Case 3**

Setup

* Manually create the /liquibase_libs directory as a peer to /lib
* Copy a driver into liquibase_libs
* Do not specify the driver in the classpath.

_Verify Liquibase finds the database driver when the driver is in the /liquibase_libs directory and the driver is not specified in the classpath._

* liquibase update --changelog-file <mychangelog> is successful.

**Case 4**

Setup

* Remove driver from /lib directory
* Make sure that driver is present in /liquibase_libs directory
* Do not specify the driver in the classpath.

_Verify Liquibase finds the database driver when the driver is in the /liquibase_libs and not present in /lib directory, and the driver is not specified in the classpath._

* liquibase update --changelog-file <mychangelog> is successful.

**Automated Tests**

No new automated functional tests are required for this change.



┆Issue is synchronized with this [Jira Bug](https://datical.atlassian.net/browse/LB-2015) by [Unito](https://www.unito.io)
